jax/numpy/lax_numpy.py

```diff
@@ -2277,11 +2277,18 @@ def atleast_3d(*arys):
 def array(object, dtype=None, copy=True, order="K", ndmin=0):
   if order is not None and order != "K":
     raise NotImplementedError("Only implemented for order='K'")
+  # check if the given dtype is compatible with JAX
   lax._check_user_dtype_supported(dtype, "array")
   dtype = dtype and dtypes.canonicalize_dtype(dtype)
 
   if _can_call_numpy_array(object):
     object = _np_array(object, dtype=dtype, ndmin=ndmin)
+
+  if type(object) is np.ndarray:
+    # check if the inferred np type is compatible with JAX
+    _inferred_dtype = object.dtype and dtypes.canonicalize_dtype(object.dtype)
+    lax._check_user_dtype_supported(_inferred_dtype, "array")
+
```

Comment:
I think  the check should be put in the existing `if type(object) is np.ndarray` block just below this.

Reply:
Actually, the problem appears when a dtype is inferred by numpy at line 2285 (on the new object). I believe that the check should be after that point (that's the path in the current codebase that is not covered - a numpy array with the dtype=object).

Reply:
Understood - but another path to this is if the input is a numpy array with dtype `object`, and it was not clear to me without looking at the source of `_can_call_numpy_array` whether that block would be executed in that case. It turns out it is.  But due to that obfuscation I thought it would be clearer to put the check in the place where it's obvious at first glance that it will be called for *every* numpy array input.

What do you think?

Reply:
I understand your point, nice catch, thanks. The check flow now should be: check current object dtype -> check specified dtype -> check inferred dtype. I don't think we can move it to the top, because in the case when the dtype is `None` and we can't infer the type from the object (like in the test case), we gen the dtype after the numpy array is being created and we check that. You are right that a check should be at the top as well in the case when a numpy array is being passed with a bad dtype to not propagate the error until the constructor of numpy array is called.

What do you think?

Reply:
The new version seems redundant: if the input is a numpy array, the typecheck now happens twice. What I had in mind is this:
```python
def array(object, dtype=None, copy=True, order="K", ndmin=0):
  if order is not None and order != "K":
    raise NotImplementedError("Only implemented for order='K'")

  # check if the given dtype is compatible with JAX
  lax._check_user_dtype_supported(dtype, "array")
  dtype = dtype and dtypes.canonicalize_dtype(dtype)

  if _can_call_numpy_array(object):
    object = _np_array(object, dtype=dtype, ndmin=ndmin)

  assert type(object) not in dtypes.python_scalar_dtypes

  if type(object) is np.ndarray:
    _inferred_dtype = object.dtype and dtypes.canonicalize_dtype(object.dtype)
    lax._check_user_dtype_supported(_inferred_dtype, "array")
    out = _device_put_raw(object)
    if dtype: assert _dtype(out) == dtype
  elif isinstance(object, (DeviceArray, core.Tracer)):
    #...
```
a single check that takes care of every case where `object` is an array.

Reply:
Sorry for being painfully stupid, I did not observe that the next block checked exactly for that, now it totally makes sense.

Reply:
No worries!